### ISSUE TITLE
Mamba2: SSM State Caching for Efficient Incremental Generation

### ISSUE THREAD

**Issue by CosmoNaught** [jax-ml/bonsai#120]
# Mamba2: SSM State Caching for Efficient Incremental Generation

## Motivation

Currently, `generate.py` recomputes the full sequence at every decoding step, giving O(n²) complexity. With state caching, we only need one forward pass per new token: O(n).

We can think of this as analogous to KV-caching in Transformers, but SSM state has a key advantage. KV-cache grows with sequence length. SSM state is fixed-size per layer: `(batch, heads, head_dim, state_size)`. The entire sequence history is compressed into this state, regardless of how long the sequence is.

## Current State

Already exists:
- `ssd_forward` accepts `initial_states` and `return_final_states`
- `Mamba2Model` threads per-layer states through

Missing:
- **Conv1d state caching**: Need to cache last `kernel_size - 1` values of `xBC` (post-projection, pre-conv). These get prepended before conv instead of zero-padding.
- **Cache object**: Unified structure holding SSM + conv states across layers
- **Generation loop**: Update to use cache instead of full recomputation

## Proposed Design

```python
@dataclasses.dataclass
class Mamba2Cache:
    ssm_states: list[jnp.ndarray]   # (batch, heads, head_dim, state_size) per layer
    conv_states: list[jnp.ndarray]  # (batch, d_conv_input, kernel_size - 1) per layer
                                    # where d_conv_input = intermediate_size + 2*state_size
```

Mixer signature becomes:
```python
def __call__(
    self, 
    hidden_states: jnp.ndarray, 
    cache: Mamba2Cache | None = None,
) -> tuple[jnp.ndarray, Mamba2Cache]:
```

Always returns `(output, cache)`. Pass `cache=None` on first call (prefill), then pass the returned cache on subsequent calls (decode).

## Tests

| Test | What it verifies |
|------|------------------|
| `test_cached_matches_full` | Process `[a,b,c]` full-sequence vs `[a,b]` then `[c]` with cache — logits for `c` must match |
| `test_conv_state_shapes` | Conv state is `(batch, intermediate_size + 2*state_size, kernel_size - 1)` |
| `test_ssm_state_shapes` | SSM state is `(batch, heads, head_dim, state_size)` per layer |
